### PR TITLE
Implement tag matching for tasks

### PR DESCRIPTION
**Pull request by gammazero**
- Tasks may have a tag that designates them for certain dealbots.
- Dealbots may be configured with 0 or more tags to request tasks by.
- Tasks will to tag match any request
- Requests that specify no tags can get any task

### CODE REVIEW

File: commands/flags.go [hunk 1]
```diff
@@ -116,6 +116,12 @@ var DaemonFlags = append(DealFlags, append(CommonFlags, append(EndpointFlags, []
 		EnvVars: []string{"DEALBOT_WORKERS"},
 		Value:   1,
 	}),
+	altsrc.NewIntFlag(&cli.IntFlag{
+		Name:    "tags",
+		Usage:   "comma separated tag strings",
+		EnvVars: []string{"DEALBOT_TAGS"},
+		Value:   1,
```
Comment (willscott) [inline commands/flags.go:123]: should this be a string array flag?

**Review by willscott — COMMENTED** [review]
looks reasonable

File: commands/flags.go (hunk 1 above)
Comment (gammazero) [reply commands/flags.go:123]: Probably, but I did not know how environ vars work with that.  Will an environ var like "tag1, tag2" be converted to ["tag1", "tag2"] ?

Comment (willscott) [reply commands/flags.go:123]: i think you want https://github.com/urfave/cli/blob/master/flag_string_slice.go#L72 "StringSliceFlag"
and yes, it should auto do the splitting of commas then

Comment (gammazero) [reply commands/flags.go]: Yes, just found out that was there.  Thank you.

File: controller/state/statedb.go [hunk 2]
```diff
@@ -277,20 +279,42 @@ func (s *stateDB) AssignTask(ctx context.Context, req tasks.PopTask) (tasks.Task
 		return nil, fmt.Errorf("cannot assign %q status to task", req.Status.String())
 	}
 
+	var tags []interface{}
+	if req.Tags.Exists() {
+		for _, t := range strings.Split(req.Tags.Must().String(), ",") {
```
Comment (mvdan) [inline controller/state/statedb.go:284]: why not make this a list of strings in IPLD directly?

File: tasks/gen.go [hunk 3]
```diff
@@ -138,6 +140,7 @@ func main() {
 	ts.Accumulate(schema.SpawnStruct("PopTask", []schema.StructField{
 		schema.SpawnStructField("Status", "Status", false, false),
 		schema.SpawnStructField("WorkedBy", "String", false, false),
+		schema.SpawnStructField("Tags", "String", true, false),
```
Comment (mvdan) [inline tasks/gen.go:143]: I think this should be called AnyTags or AllTags, to clarify if this is "at least one tag is shared with the task to pop" or "all tags must be shared with the task to pop".

**Review by mvdan — APPROVED** [review]
SGTM. I think you should make the matching semantics clearer though, because it's not terribly obvious to me from reading the code.

I think we probably want "all tags in a PopTask must be contained in the Tags list of the task being popped", i.e. also allowing extra tags to be present in the task.

File: tasks/gen.go (hunk 3 above)
Comment (mvdan) [reply tasks/gen.go:143]: Oh, and assuming this is a list too, I'd also make it an actual IPLD list.

Comment (gammazero) [reply tasks/gen.go:143]: Done

File: controller/state/statedb.go (hunk 2 above)
Comment (gammazero) [reply controller/state/statedb.go:284]: Done

**Comment by gammazero** [comment]
@mvdan Added comments to describe tag semantics.

**Review by willscott — APPROVED** [review]

**Pull request merged**